sycl/plugins/unified_runtime/ur/adapters/cuda/enqueue.cpp

```diff
@@ -284,9 +282,31 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunch(
   UR_ASSERT(workDim > 0, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
   UR_ASSERT(workDim < 4, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
 
+  std::vector<ur_event_handle_t> DepEvents(
+      phEventWaitList, phEventWaitList + numEventsInWaitList);
+
+  // phEventWaitList only contains events that are handed to UR by the SYCL
+  // runtime. However since UR handles memory dependencies within a context
+  // we may need to add more events to our dependent events list
+  for (auto &MemArg : hKernel->Args.MemObjArgs) {
+    if (MemArg.Mem->isBuffer()) {
+      MemArg.Mem->MemoryMigrationMutex.lock();
+      if (auto MemDepEvent =
+              ur_cast<ur_buffer_ *>(MemArg.Mem)->LastEventWritingToMemObj;
+          MemDepEvent && std::find(DepEvents.begin(), DepEvents.end(),
+                                   MemDepEvent) == DepEvents.end()) {
+        DepEvents.push_back(MemDepEvent);
```

Comment:
We can enqueue the copies on the same stream but in the case that `depEvents` are kernels on a different device we can't rely on streams. We need to use `CUEvents` instead.

Reply:
No the regular commands events are only the ones that are passed to the adapter by the RT, which may be from a device from a different context. Since we are unifying the context then the SYCL RT knows that it doesn't need to use dep events within the same context because the PI will handle these dependencies, as openCL does

Reply:
Oooh of course that makes sense, let me go through this again in a bit